RT_UPDATE/download_asfr.R

```diff
@@ -0,0 +1,363 @@
+#' function:
+#' \enumerate{
+#'   \item Finds the latest installed \code{wpp} package (must be \eqn{\ge} 2022 for single-year data),
+#'   \item Loads \code{percentASFR1dt} and \code{tfr1},
+#'   \item Constructs ASFR either as single-year rates or aggregated 5-year rates,
+#'   \item Interpolates across requested years \code{AsfrDatesIn} using \code{interp()}.
+#' }
+#'
+#' @param Asfrmat Optional matrix or data frame of user-supplied ASFR values.
+#'   If provided, the function returns the formatted ASFR data according to
+#'   \code{output}. \strong{Expected convention: rows = ages, columns = years.}
+#' @param location Character or numeric code specifying the country or region
+#'   for which to download data. Required if \code{Asfrmat} is not supplied.
+#' @param AsfrDatesIn Numeric vector of reference years for which ASFR data are
+#'   to be downloaded or aligned.
+#' @param method Character specifying interpolation method to use when ASFR data
+#'   are interpolated across years. Defaults to \code{"linear"}.
+#' @param output Character specifying desired output age structure:
+#'   \code{"single"} (single-year ages) or \code{"5-year"} (default).
+#' @param ... Additional arguments passed to the interpolation routine.
+#'
+#' @details
+#' The function expects WPP single-age percentage ASFR data (\code{percentASFR1dt})
+#' and total fertility rate (\code{tfr1}) to be available in the latest installed
+#' \code{wpp} package. For the \code{"single"} option, the function constructs
+#' single-year ASFR by multiplying percentage ASFR by TFR and interpolating to
+#' requested years. For the \code{"5-year"} option, the function first aggregates
+#' percentages to 5-year groups and then multiplies by TFR; it then interpolates
+#' group-level rates to requested years. When \code{Asfrmat} is provided by the
+#' user, it is converted to a matrix and rownames are set according to the
+#' requested \code{output}.
+#'
+#' @return
+#' A numeric matrix of ASFR values by age (rows) and year (columns).
+#'
+#' @importFrom dplyr filter select mutate group_by summarise ungroup left_join
+#' @importFrom dplyr group_nest
+#' @importFrom tidyr pivot_longer pivot_wider unnest
+#' @importFrom purrr map
+#' @importFrom tibble as_tibble column_to_rownames
+#' @importFrom readr parse_number
+#' @importFrom rlang .data
+#' #importFrom magrittr %>% 
+#'
+#' @examples
+#' \dontrun{
+#' # Download 5-year ASFR for France for three years
+#' asfr_data <- download_Asfr(location = "France",
+#'                            AsfrDatesIn = c(2000, 2005, 2010),
+#'                            output = "5-year")
+#'
+#' # Use user-supplied ASFR matrix (rows = ages, cols = years)
+#' Asfrmat <- matrix(runif(25), nrow = 5)
+#' download_Asfr(Asfrmat = Asfrmat,
+#'               AsfrDatesIn = c(2000, 2005, 2010, 2015, 2020),
+#'               output = "single")
+#' }
+#'
+#' @export
+
+
+download_Asfr <- function(Asfrmat     = NULL,
```

Comment:
include Age as argument rather than assuming age labels later. If Asfrmat is given and Age is also given then nrow() has to equal length(). If Asfrmat is given and it already has age labels see DemoTools::names2age() as a helper here.